### PR TITLE
HELP-44836: network container is getting the wrong container ID

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes

_Jira ticket:_[ HELP-44836](https://jira.mongodb.org/browse/HELP-44836)

- Fixes an issue with multiple containers in the same project: in the scenario where there was already a container for the specified region, the CREATE operation was using the ID of the first containers returned by the List endpoint instead of using the ID of the container in the same region as the Model.
- Proposes a new structure for the resource: I split each Handler operation into its own file to improve readability. Want to see what the team think about this new structure. **Keep in mind that we cannot create a new package because we will have a circle dependency issue (Resouce package -> Operation package -> Resource package (Model)) and we cannot change the Resource package as it is autogenerated.** 
- Updates the Delete error message to ask the user to delete the Atlas Cluster before trying to delete the container

## Type of change:

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] This change requires a documentation update

## Manual QA performed:

- [x] cfn invoke for each of CRUDL/cfn test
- [x] Updated resource in  [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples)
- [x] Published to AWS private registry
- [x] Used the template in [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples) to create and update a stack in AWS
- [x] Deleted stack to ensure resources are deleted
- [x] Created multiple resources in same stack
- [x] Validated in Atlas UI
- [x] Included screenshots

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code
- [x] For CFN Resources: I have released by changes in the private registry and proved by change
  works in Atlas

## Further comments

```bash
cfn test
========= 11 passed, 4 skipped, 9 deselected, 1 warning in 181.95s (0:03:01) =======
```

**Manual Test:**

Screenshot 2023-04-24 at 18 14 18

### Next steps

- Release the CFN Resource to all the regions
- Update the quickstart templates that are using the container and  cluster resources to use `DependsOn`
**CDK release is not needed since it is not a breaking change.**

### CODE REVIEW

File: cfn-resources/network-container/cmd/resource/resource.go [hunk 1]
```diff
@@ -81,24 +81,32 @@ func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler
...
-			containers, _, err2 := client.Containers.ListAll(context.Background(), *projectID, nil)
-			if err2 != nil {
+			containers, _, err := client.Containers.ListAll(context.Background(), *projectID, nil)
+			if err != nil {
 				_, _ = logger.Debugf("Error Containers.ListAll err:%v", err)
 				return handler.ProgressEvent{}, fmt.Errorf("error Containers.ListAll err:%v", err)
 			}
-			_, _ = logger.Debugf("containers:%v", containers)
-			if containers != nil {
-				first := containers[0]
```
Comment (andreaangiolillo) [inline cfn-resources/network-container/cmd/resource/resource.go:91]: issue - Create Operation: if a container in the same region already exists, we were returning the first container returned by `ListAll` instead of returning the ID of the container in the same region.

File: cfn-resources/network-container/cmd/resource/resource.go [hunk 2]
```diff
@@ -206,9 +214,16 @@ func Delete(req handler.Request, prevModel *Model, currentModel *Model) (handler
...
 	containerID := *currentModel.Id
 
-	response, err := deleteContainer(client, projectID, containerID)
-	if err != nil {
-		return progressevents.GetFailedEventByResponse(fmt.Sprintf("Error getting resource : %s", err.Error()),
+	if response, err := client.Containers.Delete(context.Background(), projectID, containerID); err != nil {
+		if response.StatusCode == 409 {
+			return handler.ProgressEvent{
+				OperationStatus:  handler.Failed,
+				Message:          fmt.Sprintf("Please, make sure to delete the network peering and the atlas cluster before deleting the container: %s", err.Error()),
```
Comment (andreaangiolillo) [inline cfn-resources/network-container/cmd/resource/resource.go:221]: Issue Delete Operation: you cannot delete a network container if the associated network peering is not deleted and the atlas cluster in the same region is not deleted

File: cfn-resources/network-peering/mongodb-atlas-networkpeering.json [hunk 3]
```diff
@@ -1,95 +1,97 @@
...
+  "required": [
+    "ProjectId",
+    "ContainerId",
+    "VpcId"
+  ],
+  "createOnlyProperties": [
+    "/properties/AwsAccountId",
+    "/properties/VpcId",
+    "/properties/Profile",
+    "/properties/ProjectId"
```
Comment (andreaangiolillo) [inline cfn-resources/network-peering/mongodb-atlas-networkpeering.json:62]: Drive-by fix: The network peering did not have profile and projectId as createOnly Prop

File: examples/network-container/container-peering-cluster.yml [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    Description: "Secret Manager Profile that contains the Atlas Programmatic keys."
+  ProjectId:
+    Type: String
+    Description: "The ID of your atlas project."
+
+Mappings: {}
+Resources:
+  Cluster:
+    Type: 'MongoDB::Atlas::Cluster'
+    DependsOn: NetworkContainer # Important to wait for the Network Container, otherwise the network container deletion will fail when deleting the stack
```
Comment (andreaangiolillo) [inline examples/network-container/container-peering-cluster.yml:51]: added example with cluster

File: cfn-resources/network-container/cmd/resource/create.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	container, httpResponse, err := client.Containers.Create(context.Background(), projectID, request)
+	if err == nil {
+		return container.ID, nil
+	}
+
+	if httpResponse.StatusCode != http.StatusConflict {
+		return "", fmt.Errorf("error creating network container: %w", err)
+	}
+
+	_, _ = logger.Debugf("Container already exists for this group. Try return existing container. err: %v", err)
```
Comment (colm-quinn) [inline cfn-resources/network-container/cmd/resource/create.go:85]: [q] What happens to debug log messages - can we see them in the CFN console or CloudWatch?

Goal here is to make sure we get access too all logs when debugging customer issues.

File: cfn-resources/network-container/cmd/resource/create.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	containers, _, err := client.Containers.ListAll(context.Background(), projectID, nil)
+	if err != nil {
+		return "", fmt.Errorf("error Containers.ListAll err:%v", err)
+	}
+
+	for i := range containers {
+		if containers[i].RegionName == request.RegionName {
+			return containers[i].ID, nil
+		}
+	}
```
Comment (colm-quinn) [inline cfn-resources/network-container/cmd/resource/create.go:95]: ❤️

**Review by colm-quinn — DISMISSED** [review]
LGTM from the file split and general coding aspect. I like individual files for each of the CRUDL operations - I think it's a strong match for how the resources work as well.

@maastha can you review this PR as well, in case you spot anything from the code side I've missed.

File: examples/network-container/container-peering-cluster.yml (hunk 4 above)
Comment (maastha) [reply examples/network-container/container-peering-cluster.yml:51]: [q] a little confused here, my understanding is basically Cluster should be deleted first and once that's done we can proceed with container deletion, is that right?
If yes, shouldn't it be that NetworkContainer resource 'dependsOn' Cluster resource?
Although I do see the important note here https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-dependson.html , wanted to confirm

**Review by maastha — DISMISSED** [review]
Just need some clarifications, otherwise LGTM. Like the new structure as well.

File: examples/network-container/container-peering-cluster.yml (hunk 4 above)
Comment (andreaangiolillo) [reply examples/network-container/container-peering-cluster.yml:51]: Using the link you provided as an example:
resource A is the cluster and resource B is the network container -> Resource A is deleted before resource B.

```
For example, if the properties of resource A use a !Ref to resource B, the following rules apply:

Resource B is created before resource A.
Resource A is deleted before resource B.
Resource B is updated before resource A.
```

File: cfn-resources/network-container/cmd/resource/list.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+		return *peErr, nil
+	}
+
+	projectID := *currentModel.ProjectId
+	containerRequest := &mongodbatlas.ContainersListOptions{
+		ProviderName: constants.AWS,
+		ListOptions:  mongodbatlas.ListOptions{},
+	}
+	_, _ = logger.Debugf("List projectId:%v, containerRequest:%v", projectID, containerRequest)
+	containerResponse, _, err := client.Containers.List(context.TODO(), projectID, containerRequest)
```
[the remark this answers is not in the document]
Comment (andreaangiolillo) [reply cfn-resources/network-container/cmd/resource/list.go]: updated

File: cfn-resources/network-container/cmd/resource/create.go (hunk 5 above)
Comment (andreaangiolillo) [reply cfn-resources/network-container/cmd/resource/create.go:85]: > What happens to debug log messages - can we see them in the CFN console or CloudWatch?

No, logs are not available when the resource is running on the lambda function as AWS does not share them.
These logs are printed only when running `sam local start-lambda`

**Review by maastha — APPROVED** [review]
LGTM

File: cfn-resources/network-container/cmd/resource/resource.go [hunk 8]
```diff
@@ -15,274 +15,35 @@
...
-
-func deleteContainer(client *mongodbatlas.Client, projectID string, containerID string) (*mongodbatlas.Response, error) {
-	response, err := client.Containers.Delete(context.Background(), projectID, containerID)
-
-	// handling "CANNOT_DELETE_RECENTLY_CREATED_CONTAINER" error
-	if err != nil && response.StatusCode == 409 {
-		time.Sleep(time.Second * 3)
-		return client.Containers.Delete(context.Background(), projectID, containerID)
-	}
-	return response, err
```
Comment (andreaangiolillo) [inline cfn-resources/network-container/cmd/resource/resource.go:229]: This is needed for the release process
